### ISSUE TITLE
di-live installation from live OS failing ("non-zero exitcode (32) for command: mount -o bind /dev /target/dev")

### ISSUE THREAD

**Issue by OnGle** [turnkeylinux/tracker#1041]
Issues in di-live currently prevent installing a live system.

----

[update by @jedmeister] To clarify, this bug still occurs with the latest `di-live` package (`di-live_0.9.5+30+g3fa70dc_amd64.deb`). To reproduce this bug with v15.0:

- boot ISO into live mode (i.e. don't install)
- manually run `di-live` (from within the live system)
- experience the failure...! :cry:

**Comment by JedMeister** [comment]
Note, whilst the nature of this issue changes depending on which build of di-live is included, all current iterations of di-live on stretch are broken... :cry: Please output for each build of di-live we've currently been testing:

Package currently in tkl stretch repo ("normal" ISO install also broken):
```
root@core ~# apt-cache policy di-live | grep Installed
  Installed: 0.9.5+25+gf7ab328
root@core ~# di-live -d
Traceback (most recent call last):
  File "/usr/lib/di-live.d/42partman-base", line 67, in <module>
    main()
  File "/usr/lib/di-live.d/42partman-base", line 61, in main
    preset_debconf(resets=templates, preseeds=preseeds, seen=seen)
  File "/usr/lib/di-live.d/common.py", line 129, in preset_debconf
    db.reset(template)
  File "/usr/lib/python2.7/dist-packages/debconf.py", line 62, in <lambda>
    lambda *args, **kw: self.command(command, *args, **kw))
  File "/usr/lib/python2.7/dist-packages/debconf.py", line 98, in command
    raise DebconfError(status, data)
debconf.DebconfError: (10, "partman-auto/init_automatically_partition doesn't exist")
Debian Installer Live
---------------------

  1. Partman-base  2. Live-installer  3. Grub-installer  4. Finalize
Debian Installer Live Menu 
```
With all reproducible changes rolled back (i.e. latest package in dev-test repo; essentially the same as what we previously had in dev-test repo & jessie - "normal" ISO install works):
```
root@core ~# apt-cache policy di-live | grep Installed
  Installed: 0.9.5+26+gc09546e
root@core ~# apt-cache policy di-live | grep Installed
  Installed: 0.9.5+26+gc09546e
root@core ~# di-live -d
Starting up the partitioner
---------------------------

..14%
..21%
..35%
..42%
..50%
..64%
..71%
..85%
..92%
..100%
Debian Installer Live
---------------------

The installer can guide you through partitioning a disk (using different standard schemes) or, if you prefer, you can do it manually. With guided partitioning you
will still have a chance later to review and customise the results.

If you choose guided partitioning for an entire disk, you will next be asked which disk should be used.

  1. Guided - use entire disk and set up LVM  2. Guided - use entire disk  3. Manual
Partitioning method: 1

Before the Logical Volume Manager can be configured, the current partitioning scheme has to be written to disk. These changes cannot be undone.

After the Logical Volume Manager is configured, no additional changes to the partitioning scheme of disks containing physical volumes are allowed during the
installation. Please decide if you are satisfied with the current partitioning scheme before continuing.

The partition tables of the following devices are changed:
 SCSI3 (0,0,0) (sda)

Write the changes to disks and configure LVM? [yes/no] yes

Guided partitioning
-------------------

..20%
..40%
You may use the whole volume group for guided partitioning, or part of it. If you use only part of it, or if you add more disks later, then you will be able to grow
logical volumes later using the LVM tools, so using a smaller part of the volume group at installation time may offer more flexibility.

The minimum size of the selected partitioning recipe is 1.4 GB (or 66%); please note that the packages you choose to install may require more space than this. The
maximum available size is 2.1 GB.

Hint: "max" can be used as a shortcut to specify the maximum size, or enter a percentage (e.g. "20%") to use that percentage of the maximum size.

Amount of volume group to use for guided partitioning: 90%

..60%
..80%
..100%
Starting up the partitioner
---------------------------

..14%
..21%
..35%
..42%
..50%
..64%
..71%
..85%
..92%
..100%
Debian Installer Live
---------------------

If you continue, the changes listed below will be written to the disks. Otherwise, you will be able to make further changes manually.

The partition tables of the following devices are changed:
 LVM VG turnkey, LV root
 LVM VG turnkey, LV swap_1

The following partitions are going to be formatted:
 LVM VG turnkey, LV root as ext4
 LVM VG turnkey, LV swap_1 as swap

Write the changes to disks? [yes/no] yes

Partitions formatting
---------------------

..33%
Partitions formatting
---------------------

..33%
Installing the system...
------------------------

  1. Partition disks  2. Install the system  3. Install the GRUB boot loader on a hard disk  4. Finalize
Debian Installer Live Menu 2

Installing the system...
------------------------

  1. Partition disks  2. Install the system  3. Install the GRUB boot loader on a hard disk  4. Finalize
Debian Installer Live Menu 3

Traceback (most recent call last):
  File "/usr/lib/di-live.d/73grub-installer", line 41, in <module>
    main()
  File "/usr/lib/di-live.d/73grub-installer", line 35, in main
    targetmounts = TargetMounts('/target')
  File "/usr/lib/di-live.d/73grub-installer", line 14, in __init__
    self.mount()
  File "/usr/lib/di-live.d/73grub-installer", line 18, in mount
    system("mount -o bind /dev %s" % self.targetdev)
  File "/usr/lib/di-live.d/common.py", line 108, in system
    raise ExecError(command, exitcode)
common.ExecError: non-zero exitcode (32) for command: mount -o bind /dev /target/dev
Exception common.ExecError: ExecError('umount -f /target/dev', 32) in <bound method TargetMounts.__del__ of <__main__.TargetMounts instance at 0x7f9465f20d40>> ignored
  1. Partition disks  2. Install the system  3. Install the GRUB boot loader on a hard disk  4. Finalize
Debian Installer Live Menu 
```

Latest dev-test package (i.e. includes [this commit](https://github.com/JedMeister/di-live/commit/6f61633a4af1ee96af4e25f048c246dfdf92ffe1) instead of removing the reproducible packaging stuff - "normal" ISO install works again):
```
root@core ~# apt-cache policy di-live | grep Installed
  Installed: 0.9.5+26+g6f61633
root@core ~# di-live -d
Starting up the partitioner
---------------------------

..14%
..21%
..35%

[...] same as above...
```

**Comment by lamby** [comment]
Just braindumping for now but if one builds a package pre- and post-reproducibility changes you can see:

 * A bunch of debconf template translations different  in the diff
 * `I: di-live: unused-debconf-template partman-auto/init_automatically_partition` [Lintian](https://lintian.debian.org/) messages amongst other `unused-debconf-template`.

(My gut/spidey sense is this is due to some of the bigger changes to `debian/rules` and possibility in the upgrade of the compat level resulting in different behaviour of, for example, `dh_installdebconf` or similar.

Will look into this more.

**Comment by JedMeister** [comment]
@lamby - To clarify, this issue is related to launching the ISO into live mode. Then from the live OS, attempting to install onto disk. Personally I would call that a "live install", but perhaps our terminology differs?

Also, I woke up this morning with a feeling I should double check what happens on v14.2. Turns out that this bug (let's call it `non-zero exitcode (32) for command: mount -o bind /dev /target/dev`) also occurs on v14.2 but escaped our testing! Here's is the stacktrace from this same scenario in v14.2:
```
Traceback (most recent call last):
  File "/usr/lib/di-live.d/73grub-installer", line 41, in <module>
    main()
  File "/usr/lib/di-live.d/73grub-installer", line 35, in main
    targetmounts = TargetMounts('/target')
  File "/usr/lib/di-live.d/73grub-installer", line 14, in __init__
    self.mount()
  File "/usr/lib/di-live.d/73grub-installer", line 18, in mount
    system("mount -o bind /dev %s" % self.targetdev)
  File "/usr/lib/di-live.d/common.py", line 108, in system
    raise ExecError(command, exitcode)
common.ExecError: non-zero exitcode (32) for command: mount -o bind /dev /target/dev
Exception common.ExecError: ExecError('umount -f /target/dev', 32) in <bound method TargetMounts.__del__ of <__main__.TargetMounts instance at 0x7fe2c764bcb0>> ignored
```

So my reading of the situation is that this bug (i.e. `non-zero exitcode (32) for command: mount -o bind /dev /target/dev` - the stacktrace in this post) is completely unrelated to #1042. 

The bit of my first post in this thread (i.e. `partman-auto/init_automatically_partition doesn't exist`) is just a confusion, and is exactly the same issue as #1042

**Comment by lamby** [comment]
Cool, I'll jump off this bug for now then? I'm not sure I can help beyond the obvious "the `umount` is probably failing as something is still using something on `/target/dev`. :)

**Comment by JedMeister** [comment]
@qrntz - do you want to have a look at this one ASAP? 

Whilst it's far from ideal, seeing as it was included in v14.2 and I din't get any bug reports relating to it, I'm tempted to not let it block v15.0 (and push this issue to v15.1). But OTOH it would be nice if it worked in v15.0?! Thoughts?

**Comment by OnGle** [comment]
Just jumping in quickly here, I believe the issue is as follows:

1. attempt to mount bind `/dev` to targetdev in `TargetMounts.mount` fails
2. during cleanup `TargetMounts` instance's reference count drops to zero and it's `__del__` method gets called
3. in `TargetMounts.__del__` `self.umount` is called, which then attempts to unmount the previously mounted proc and dev, which fails because dev isn't mounted.

There is some unknown issue occuring in step 1 but I think its safe to say regardless of that we should be handling mount cleanup in a way that wont cause later exceptions to hide the real issues.
